### ISSUE TITLE
Possibly unnecessary stream GetString()

### ISSUE THREAD

**Issue by buchmoyerm** [connamara/quickfixn#299]
I was looking around the SocketInitiatorThread.cs when I noticed line 82 converts the byte[] to a string which is sent to the parser

```
    parser_.AddToStream(System.Text.Encoding.UTF8.GetString(readBuffer_, 0, bytesRead));
```

Then Parser.cs immediately converts that string back to a byte array

```
    public void AddToStream(string data)
    {
        byte[] bytes = System.Text.Encoding.UTF8.GetBytes(data);
        AddToStream(ref bytes, bytes.Length);

    }
```

Is there a reason why line 82 of SocketInitiatorThread.cs can't just add the byte[] to the parser directly?

```
    parser_.AddToStream(ref readBuffer_, bytesRead);
```

I have made this change in my local copy and every appears to be working correctly. Is there a reason that the buffer would need to be converted to a string before being passed to the parser?

**Comment by staffanu** [comment]
I've been looking at this as well and also submitted a pull request with a fix (https://github.com/connamara/quickfixn/pull/285). Apart from being inefficient (converting between strings and byte arrays unnecessarily) the code also breaks multi-byte characters (when they appear on the boundary between the arbitrarily broken up strings).

It was argued against the pull request that multi-byte chanracters are not allowed anyway, but actually they are in some fields (and de facto, they are used by many people, including us, in private systems). I would argue that this layer should pass bytes cleanly to the parser. Having a different byte array at the parser than captured from the network just seems wrong.

BTW, if you're interested in performance I made a fork (staffanu/quickfixn) for the changes we use in a quite high throughput system. Apart from the pull requests I've submitted already, that branch includes quite a few performance improvements: fewer conversions to/from strings and byte arrays, a circular buffer for receiving data, and also a single pass message to byte array conversion (as opposed to the standard three pass algorithm). The main reason for my changes was to reduce the amount of garbage produced (it is about half), but performance also increased considerably (but I did not really measure how much).

Staffan
